### ISSUE TITLE
Compilation error

### ISSUE THREAD

**Issue by kierepka** [fsbolero/Bolero#120]
After small changes in code I'm getting this error on MacOs:

 dotnet run -c Debug -p src/Eden.Client
/Users/mateusz/.nuget/packages/microsoft.aspnetcore.blazor.build/3.2.0-preview1.20073.1/targets/Blazor.MonoRuntime.targets(258,5): error : Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object. [/Users/mateusz/Projects/Eden/src/Eden.Client/Eden.Client.fsproj]
/Users/mateusz/.nuget/packages/microsoft.aspnetcore.blazor.build/3.2.0-preview1.20073.1/targets/Blazor.MonoRuntime.targets(258,5): error : ILLink failed with exit code 134. [/Users/mateusz/Projects/Eden/src/Eden.Client/Eden.Client.fsproj]

Standard empty bolero template is running well.

**Comment by Tarmil** [comment]
I haven't tried 3.2.0-preview1 yet. Does the same happen with 3.1.0-preview4?

**Comment by kierepka** [comment]
dotnet run
crit: Microsoft.AspNetCore.Hosting.Diagnostics[6]
      Application startup exception
System.IO.DirectoryNotFoundException: C:\Temp\eden\src\Eden.Client\bin\Debug\netstandard2.1\dist\
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root, ExclusionFilters filters)
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root)
   at Microsoft.AspNetCore.Builder.BlazorHostingApplicationBuilderExtensions.UseClientSideBlazorFiles(IApplicationBuilder app, String clientAssemblyFilePath)
   at Microsoft.AspNetCore.Blazor.DevServer.Server.Startup.Configure(IApplicationBuilder app, IWebHostEnvironment environment, IConfiguration configuration)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.Invoke(Object instance, IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.<>c__DisplayClass4_0.<Build>b__0(IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass13_0.<UseStartup>b__2(IApplicationBuilder app)
   at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
Unhandled exception. System.IO.DirectoryNotFoundException: C:\Temp\eden\src\Eden.Client\bin\Debug\netstandard2.1\dist\
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root, ExclusionFilters filters)
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider..ctor(String root)
   at Microsoft.AspNetCore.Builder.BlazorHostingApplicationBuilderExtensions.UseClientSideBlazorFiles(IApplicationBuilder app, String clientAssemblyFilePath)
   at Microsoft.AspNetCore.Blazor.DevServer.Server.Startup.Configure(IApplicationBuilder app, IWebHostEnvironment environment, IConfiguration configuration)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.Invoke(Object instance, IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.<>c__DisplayClass4_0.<Build>b__0(IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass13_0.<UseStartup>b__2(IApplicationBuilder app)
   at Microsoft.AspNetCore.HostFilteringStartupFilter.<>c__DisplayClass0_0.<Configure>b__0(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Microsoft.AspNetCore.Blazor.DevServer.Commands.ServeCommand.Execute()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Microsoft.AspNetCore.Blazor.DevServer.Program.Main(String[] args)

**Comment by Happypig375** [comment]
Please post your "small changes in code".

**Comment by brikken** [comment]
I'm getting a similar build error when trying to upgrade to 3.2.0-preview1.20073.1

**Comment by brikken** [comment]
So far, I can make it build using the new version of Blazor for the new Program.fs structure, but only with the previous version of Blazor.Build

https://github.com/brikken/FsBoleroTemplateBlazorPreview3.2/commit/cf4fe263487e663a223ea3265359a6c0458c76d2

**Comment by Tarmil** [comment]
I just tried upgrading Bolero itself and got to the same conclusion: it only builds if I leave Microsoft.AspNetCore.Build on 3.1-preview4. Looks like the linker doesn't like something about our assemblies, here's hoping it's something in Bolero that we can fix and not eg something in FSharp.Core.

**Comment by Tarmil** [comment]
Bad news, the error happens with a plain C# blazor template just by referencing FSharp.Core:

```shell
$ dotnet new blazorwasm
$ dotnet build
(Succeeds)
$ dotnet add package FSharp.Core
$ dotnet build
(Fails)
```

I reported the issue on the linker repo: https://github.com/mono/linker/issues/927

**Comment by brikken** [comment]
Ouch, this is quite a hit.

**Comment by OnurGumus** [comment]
Perhaps we can tell the linker not to trim FSharp.Core as a temporary workaround ?

**Comment by Tarmil** [comment]
Adding FSharp.Core to linker.xml doesn't seem to change anything. I'm assuming it still analyzes the assembly to link its dependencies.

Of course you can skip linking altogether with `<BlazorLinkOnBuild>false</BlazorLinkOnBuild>` if you don't mind the 18MB app...

**Comment by kierepka** [comment]
So is there any other option? For development propouses skipping linking is not so bad idea, but I would like to publish new changes ASAP ;) 

ps. https://endenczyk.pl - this is my site where I'm learning how to use bolero and F#.
Now I'm checking how to save to local storage based on https://github.com/Blazored/LocalStorage/tree/master/src/Blazored.LocalStorage

**Comment by Tarmil** [comment]
I've tested more, and I can confirm that we can keep `Microsoft.AspNetCore.Blazor.Build` on 3.1-preview4 while upgrading everything else to 3.2-preview1. It works nicely and the full Bolero test suite passes.

**Comment by kierepka** [comment]
After changing References:
    <PackageReference Include="Bolero" Version="0.11.22-preview31" />
    <PackageReference Include="Bolero.Build" Version="0.11.22-preview31" />
    <PackageReference Include="Bolero.HotReload" Version="0.11.5-preview31" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.Build" Version="3.1-preview4.19579.2" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.DevServer" Version="3.1-preview4.19579.2" />

I'm sure that problem is in Bolero.Templating.dll:

    /Users/mateusz/.nuget/packages/microsoft.aspnetcore.blazor.build/3.1.0-preview4.19579.2/targets/Blazor.MonoRuntime.targets(204,5): error : Unhandled exception. Mono.Linker.MarkException: Error processing method: 'Microsoft.FSharp.Control.FSharpAsync`1<Microsoft.FSharp.Core.Unit> Bolero.Templating.Client.ClientBase::RefreshAllFiles()' in assembly: 'Bolero.Templating.dll'
     ---> Mono.Cecil.ResolutionException: Failed to resolve Microsoft.FSharp.Control.FSharpAsync`1<!!0[]> Microsoft.FSharp.Control.FSharpAsync::Parallel(System.Collections.Generic.IEnumerable`1<Microsoft.FSharp.Control.FSharpAsync`1<!!0>>,Microsoft.FSharp.Core.FSharpOption`1<System.Int32>)
       at Mono.Linker.Steps.MarkStep.HandleUnresolvedMethod(MethodReference reference)
       at Mono.Linker.Steps.MarkStep.MarkMethod(MethodReference reference)
       at Mono.Linker.Steps.MarkStep.MarkInstruction(Instruction instruction)
       at Mono.Linker.Steps.MarkStep.MarkMethodBody(MethodBody body)
       at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method)
       at Mono.Linker.Steps.MarkStep.ProcessQueue()
       --- End of inner exception stack trace ---
       at Mono.Linker.Steps.MarkStep.ProcessQueue()
       at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue()
       at Mono.Linker.Steps.MarkStep.Process()
       at Mono.Linker.Steps.MarkStep.Process(LinkContext context)
       at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step)
       at Mono.Linker.Pipeline.Process(LinkContext context)
       at Mono.Linker.Driver.Run(ILogger customLogger)
       at Mono.Linker.Driver.Execute(String[] args, ILogger customLogger)
       at Mono.Linker.Driver.Main(String[] args)
    /Users/mateusz/.nuget/packages/microsoft.aspnetcore.blazor.build/3.1.0-preview4.19579.2/targets/Blazor.MonoRuntime.targets(204,5): error : ILLink failed with exited code 134.
Done building target "_LinkBlazorApplication" in project "Eden.Client.fsproj" -- FAILED.

**Comment by Tarmil** [comment]
@kierepka This is a different problem. The NullReferenceException you posted first is definitely a problem with FSharp.Core: as posted in mono/linker#927, simply creating a C# Blazor app and referencing FSharp.Core causes it.

Would you mind posting this one in a new issue? Thanks.

**Comment by Tarmil** [comment]
This issue is fixed in Blazor 3.2-preview2, so it will be fixed in Bolero as soon as we upgrade to it (#125).

**Comment by Tarmil** [comment]
#125 merged, the next release will be on Blazor 3.2.
